eclass/databases.eclass

```diff
@@ -0,0 +1,429 @@
+
+case ${EAPI} in
+	8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI} unsupported."
+esac
+
+if [[ ! ${_DATABASES_ECLASS} ]]; then
+_DATABASES_ECLASS=1
+
+# ==============================================================================
+# GENERIC FUNCTIONS
+# ==============================================================================
+
+# @FUNCTION: _databases_gen_depend
+# @USAGE: <funcname>
+# @DESCRIPTION:
+# Get a dependency string for the given helper function.
+_databases_gen_depend() {
+	local srvname=${1:1}
+	case ${srvname} in
+		memcached)
+			echo "net-misc/memcached"
+			;;
+		mongod)
+			echo "dev-db/mongodb"
+			;;
+		mysql)
+			echo "virtual/mysql[server]"
+			;;
+		postgresql)
+			echo "dev-db/postgresql[server]"
+			;;
+		redis)
+			echo "dev-db/redis"
+			;;
+		*)
+			die "${ECLASS}: unknown database: ${srvname}"
+	esac
+}
+
+# @FUNCTION: _databases_stop_service
+# @USAGE: <funcname>
+# @DESCRIPTION:
+# Default function to stop servers.  Reads PID from a file and sends the TERM
+# signal.
+_databases_stop_service() {
+	local funcname=${1?}
+	local srvname=${funcname:1}
+	local pidfile="$(${funcname} get-pidfile)"
+
+	ebegin "Stopping ${srvname}"
+	kill "$(<"${pidfile}")"
+	eend $? || die -n
+}
+
+# @FUNCTION: _databases_dispatch
+# @USAGE: <funcname> <cmd> [port]
+# @INTERNAL
+# @DESCRIPTION:
+# Start or stop a service.
+_databases_dispatch() {
+	local funcname=${1?}
+	local cmd=${2?}
+	local port=${3}
+
+	local start_fn=( _${funcname}_start ${port} )
+	if ! declare -f "${start_fn[0]}" >/dev/null; then
+		die "${ECLASS}: function not declared: ${start_fn[0]}"
+	fi
+
+	local stop_fn=( _databases_stop_service ${funcname} )
+	if declare -f _${funcname}_stop >/dev/null; then
+		stop_fn=( _${funcname}_stop )
+	fi
+
+	case ${cmd} in
+		get-depend)
+			_databases_gen_depend ${funcname}
+			;;
+		get-dbpath)
+			echo "${T}"/db/${funcname}/
+			;;
+		get-logfile)
+			echo "${T}"/db_log/${funcname}
+			;;
+		get-pidfile)
+			echo "${T}"/run/${funcname}.pid
+			;;
+		get-sockdir)
+			echo "${T}"/run/${funcname}/
+			;;
+		get-sockfile)
+			echo "${T}"/run/${funcname}.sock
+			;;
+		--start)
+			debug-print-function "${start_fn[@]}"
+			"${start_fn[@]}"
+			;;
+		--stop)
+			debug-print-function "${stop_fn[@]}"
+			"${stop_fn[@]}"
+			;;
+		*) die "${funcname}: invalid command: ${cmd}" ;;
+	esac
+}
+
+# ==============================================================================
+# MEMCACHED
+# ==============================================================================
+
+# @FUNCTION: _ememcached_start
+# @USAGE: [port]
+# @INTERNAL
+# @DESCRIPTION:
+# Start memcached server.
+_ememcached_start() {
+	local port=${1:-11211}
+	local pidfile=$(ememcached get-pidfile)
+	mkdir -p "$(dirname "${pidfile}")" || die
+
+	local myargs=(
+		--daemon
+		--port=${port}
+		--user=nobody
+		--listen=127.0.0.1
+		--pidfile="${pidfile}"
+	)
+
+	ebegin "Spawning memcached"
+	memcached "${myargs[@]}" > /dev/null
+	eend $? || die -n
```

Comment:
The `die` line here doesn't provide enough context to tell what failed, so you need a specific message.